### DISCUSSION TITLE
[Repost with additional info] Geocentric altitude/az of the sun and moon.

### DISCUSSION THREAD

**Discussion by hidp123** [skyfielders/python-skyfield#860] [Q&A] [answered]
Sorry to repost this, I have added additional info which I hope will help in answering this query, and thus give a more accurate value.
___
Normally, alt & az are topocentric. However, how can I get geocentric alt & az. The altitude at least, if not azimuth.

I ask mainly due to the following (screenshot taken from [this](https://www.astronomycenter.net/pdf/yallop_1997.pdf) paper by Bernard Yallop 1997):

![image](https://user-images.githubusercontent.com/84683703/231354076-5c011bb8-0c41-4c4b-83c2-bbea4e12f8ee.png)

So far looking at other libraries I have found that Swiss Ephemeris has this option, and cosinekitty at [astronomy-engine](https://github.com/cosinekitty/astronomy) managed to put together something for me in this regard using his library. [Here](https://github.com/cosinekitty/astronomy/discussions/299#discussioncomment-5587004) (Swiss Ephemeris) and [here](https://github.com/cosinekitty/astronomy/discussions/299#discussioncomment-5598927) is my discussion with cosinekitty in regards to this, hopefully it helps us here as well.
I was hoping it can also be calculated via skyfield due to skyfield's high level of accuracy.

Thanks

**Comment by brandon-rhodes** [comment, marked as answer]
~I can't make any sense of the phrase that defines _ARCV_, since there's no altitude from the center of the Earth, because there's no horizon there. Have you tried creating a lat/lon whose altitude is negative enough to put it at, or near, the center of the Earth? You could try something like elevation -6378 meters and see if the numbers agree with the paper.~

**Edit:** This weekend I had time to go back and read the crescent-Moon literature a bit more carefully, and I see that I was wrong about how I was trying to interpret the phrase "geocentric altitude". Instead, it really does mean "altitude" in the sense of "altitude-azimuth", computed as though the observer _and_ their local horizon were somehow located down at the center of the Earth.

So I've figured out how Skyfield can compute it! The full details are here, on issue #951, which also asked about that quantity:

https://github.com/skyfielders/python-skyfield/issues/951#issuecomment-2018546762

(Edited to remove wrong version of script; let's instead just link to #951, so the script can be updated there, in one place only, if further bugs are found.)

**Reply by Juned-Patel** [reply]
Yes, we typically use topocentric altitudes for practical purposes, but for many astronomical calculations, we require geocentric altitudes. Is it possible to include this feature? For instance, in the example provided, the distance between Mars and Earth can be calculated using Earth's geocentric and topocentric positions. Similarly, we would like to determine the geocentric position of altitude and azimuth.

https://rhodesmill.org/skyfield/

from skyfield.api import load

# Create a timescale and ask the current time.
ts = load.timescale()
t = ts.now()

# Load the JPL ephemeris DE421 (covers 1900-2050).
planets = load('de421.bsp')
earth, mars = planets['earth'], planets['mars']

# What's the position of Mars, viewed from Earth?
astrometric = earth.at(t).observe(mars)
ra, dec, distance = astrometric.radec()

print(ra)
print(dec)
print(distance)

**Reply by brandon-rhodes** [reply]
Could you provide an example of the angles and distance that you would like returned for a specific date and time? That will be necessary to confirm whether Skyfield can compute those values or not.

**Reply by hidp123** [reply]
**Results:**
sun_**az** _ **geo**: 252.572090375034
sun_**az** _ **topo**: 252.57208136484496

sun_**alt** _ noRefrac_**geo**: 8.57189942754521
sun_**alt** _ noRefrac_**topo**: 8.569480940769978

moon_**alt** _ noRefrac_**geo**: 11.687140086069506
moon_**alt** _ noRefrac_**topo**: 10.68370977407998
Bigger difference for moon alt compared to the sun.

What I have noticed is that swisseph converts from ecliptic longitude and latitude to Alt/Az.
So, for geocentric it uses the geocentric ecl. long. lat., and for topocentric it uses topocentric ecl. long. lat, then it converts that to Alt/Az.

Code from Pyswisseph below to get the above results (for the sun):
**For geocentric alt/az:**

```
import swisseph as swe

# convert date and time to Julian Days
jd_UT = swe.julday(year = 2024, month = 3, day = 17, hour = 0, cal = swe.GREG_CAL)


""" GEO """
# sun apparent geocentric ecliptic long, lat, dist:
sun_geo_ecliptic_long_lat_dist_speed = swe.calc_ut(tjdut = jd_UT, planet = swe.SUN, flags = swe.FLG_SWIEPH)

sun_geo_ecliptic_long = sun_geo_ecliptic_long_lat_dist_speed[0][0]
sun_geo_ecliptic_lat = sun_geo_ecliptic_long_lat_dist_speed[0][1]

"""AZ, ALT with and without refraction (from ecl. coords):"""
sun_az_geo = swe.azalt(tjdut = jd_UT, flag = swe.ECL2HOR, geopos = (longitude, latitude, altitude), atpress = atmospheric_pressure, attemp = atmospheric_temp, xin = (sun_geo_ecliptic_long, sun_geo_ecliptic_lat, 0))[0]
if sun_az_geo < 180:      # swisseph gives az from South instead of from North. So 0° will be 180°.
    sun_az_geo += 180
else:
    sun_az_geo -= 180

sun_alt_noRefrac_geo = swe.azalt(tjdut = jd_UT, flag = swe.ECL2HOR, geopos = (longitude, latitude, altitude), atpress = atmospheric_pressure, attemp = atmospheric_temp, xin = (sun_geo_ecliptic_long, sun_geo_ecliptic_lat, 0))[1]

print("sun_az_geo:", sun_az_geo)
print("sun_alt_noRefrac_geo:", sun_alt_noRefrac_geo)

```


**For topocentric alt/az:**

```
import swisseph as swe

# convert date and time to Julian Days
jd_UT = swe.julday(year = 2024, month = 3, day = 10, hour = 17, cal = swe.GREG_CAL)


""" TOPO """
longitude = -1.5
latitude = 52
altitude = 0.0
atmospheric_pressure = 1010 # atmospheric pressure in mbar/hPa
atmospheric_temp = 10 # atmospheric temperature in degrees Celsius

swe.set_topo(lon = longitude, lat = latitude, alt = altitude) # alt in meters

# sun apparent topocentric ecliptic long, lat, dist:
sun_topo_ecliptic_long_lat_dist_speed = swe.calc_ut(tjdut = jd_UT, planet = swe.SUN, flags = swe.FLG_SWIEPH | **swe.FLG_TOPOCTR**)

sun_topo_ecliptic_long = sun_topo_ecliptic_long_lat_dist_speed[0][0]
sun_topo_ecliptic_lat = sun_topo_ecliptic_long_lat_dist_speed[0][1]

"""AZ, ALT with and without refraction (from ecl. coords):"""
sun_az_topo = swe.azalt(tjdut = jd_UT, flag = swe.ECL2HOR, geopos = (longitude, latitude, altitude), atpress = atmospheric_pressure, attemp = atmospheric_temp, xin = (sun_topo_ecliptic_long, sun_topo_ecliptic_lat, 0))[0]
if sun_az_topo < 180:
    sun_az_topo += 180
else:
    sun_az_topo -= 180

sun_alt_noRefrac_topo = swe.azalt(tjdut = jd_UT, flag = swe.ECL2HOR, geopos = (longitude, latitude, altitude), atpress = atmospheric_pressure, attemp = atmospheric_temp, xin = (sun_topo_ecliptic_long, sun_topo_ecliptic_lat, 0))[1]

print("sun_az_topo:", sun_az_topo)
print("sun_alt_noRefrac_topo:", sun_alt_noRefrac_topo)
```

**Reply by brandon-rhodes** [reply]
Thank you for the examples of code! This makes it much clearer what you mean.

I think that the answer is that your first example script is, alas, incorrect? You cannot take a geocentric pair of coordinates, which measure from the Earth's center, and turn them into an altitude and azimuth that (supposedly) measure from a geographic position. The resulting coordinates will be wrong, and a telescope pointed in that direction won't see the target.

Could you explain what overall problem you are trying to solve? That might help untangle how you wound up with this series of calls to the other astronomy library. Maybe you were trying to do something else, and wound up passing geocentric coordinates to the topocentric alt/az routine by mistake?

**Reply by hidp123** [reply]
You are right, it does feel contradictory, but why I am after this, please see my original post from previously:
https://github.com/skyfielders/python-skyfield/discussions/860#discussion-5237044
(or can scroll to the top)

**Reply by brandon-rhodes** [reply]
The paper you cited merely confused me, since it did not define the term, but it just occurred to me that I could Google the term, and see if a definition existed elsewhere. And results came up! Here is a top result with two diagrams, one in the comment linked, and the other in the comment that follows:

https://github.com/JSBSim-Team/jsbsim/issues/184#issuecomment-663856247

It looks like "geocentric altitude" is not an angle "altitude" but a distance "altitude", which makes far more sense than how I was initially trying to understand the term! Happily, there is nothing contradictory about trying to measure an altitude above the Earth's surface from the geocenter.

But I'll have to think about whether there's a good way in Skyfield to take the line between the geocenter and an object like the Moon, and compute the point on the Earth ellipsoid's surface that's on that line. I suspect there's a geometry routine in Skyfield we can use, but I'll have to do a search later to learn its name and how it's called. 

In any case, you can now return to your work with the other astronomy library, and rework it to use the concept of altitude-as-a-distance, rather than the idea of altitude-as-an-angle-above-the-horizon. It's so unfortunate that English uses the same term "altitude" for both an angle and a distance!

**Reply by brandon-rhodes** [reply]
My most recent comment, above, was wrong! Scroll up to see that I have edited my answer at the top of this thread, and have provided a sample script that I think computes the value you are after.

**Reply by Juned-Patel** [reply]
Yes, thank you for investigating the issue. I've just tested the code and confirmed that the result is accurate.

**Reply by brandon-rhodes** [reply]
Wonderful! I'll mark this as the answer, then.

Note that there was a small bug with the first version of the script that I pasted here, so I've updated the answer to simply include a link to the issue, so that if we make any further improvements, this Discussion won't be left with an out-of-date answer.

I'm glad we got your use case worked out!
